pyhf/__init__.py

```diff
@@ -307,53 +308,75 @@ def generate_asimov_data(asimov_mu, data, pdf, init_pars, par_bounds):
 def loglambdav(pars, data, pdf):
     return -2 * pdf.logpdf(pars, data)
 
+
 def qmu(mu, data, pdf, init_pars, par_bounds):
-    # The Test Statistic
-    mubhathat = tensorlib.tolist(optimizer.constrained_bestfit(loglambdav, mu, data, pdf, init_pars, par_bounds))
-    muhatbhat = tensorlib.tolist(optimizer.unconstrained_bestfit(loglambdav, data, pdf, init_pars, par_bounds))
-    qmu = tensorlib.tolist(loglambdav(mubhathat, data, pdf) - loglambdav(muhatbhat, data, pdf))[0]
-    if muhatbhat[pdf.config.poi_index] > mu:
-        return 0.0
+    """
+    The test statistic, q_mu, for establishing an upper
+    limit on the strength parameter, mu, as defiend in
+    Equation (14) in arXiv:1007.1727
+
+    .. math::
+       :nowrap:
+
+       q_{\mu} = \left\{\begin{array}{ll}
+       -2\ln\lambda\left(\mu\right), &\hat{\mu} < \mu,\\
+       0, & \hat{\mu} > \mu
+       \end{array}\right.
+
+    Args:
+        mu (Number or Tensor): The signal strength parameter
+        data (Tensor): The data to be considered
+        pdf (Tensor): The model used in the likelihood ratio calculation
+        init_pars (Tensor): The initial parameters
+        par_bounds(Tensor): The bounds on the paramter values
+
+    Returns:
+        Float: The calculated test statistic, q_mu
+    """
+    mubhathat = optimizer.constrained_bestfit(
+        loglambdav, mu, data, pdf, init_pars, par_bounds)
+    muhatbhat = optimizer.unconstrained_bestfit(
+        loglambdav, data, pdf, init_pars, par_bounds)
+    qmu = loglambdav(mubhathat, data, pdf) - loglambdav(muhatbhat, data, pdf)
+    qmu = tensorlib.where(muhatbhat[pdf.config.poi_index] > mu, [0], qmu)
     return qmu
 
-from scipy.stats import norm
+
 def pvals_from_teststat(sqrtqmu_v, sqrtqmuA_v):
     # these pvals are from formula
     # (59) in arxiv:1007.1727 p_mu = 1-F(q_mu|mu') = 1- \Phi(q_mu - (mu-mu')/sigma)
     # and  (mu-mu')/sigma = sqrt(Lambda)= sqrt(q_mu_A)
-    CLsb = 1 - norm.cdf(sqrtqmu_v)
-    CLb =  1 - norm.cdf(sqrtqmu_v - sqrtqmuA_v)
+    CLsb = 1 - tensorlib.normal_cdf(sqrtqmu_v)
+    CLb = 1 - tensorlib.normal_cdf(sqrtqmu_v - sqrtqmuA_v)
     oneOverCLs = CLb / CLsb
     return CLsb, CLb, oneOverCLs
 
-import math
+
 def runOnePoint(muTest, data, pdf, init_pars, par_bounds):
-    asimov_mu = 0.0
-    asimov_data = tensorlib.tolist(generate_asimov_data(asimov_mu, data,
-                                   pdf, init_pars, par_bounds))
+    asimov_mu = 0.
+    asimov_data = generate_asimov_data(asimov_mu, data, pdf, init_pars, par_bounds)
 
-    qmu_v  = qmu(muTest, data, pdf, init_pars, par_bounds)
+    qmu_v = qmu(muTest, data, pdf, init_pars, par_bounds)
     qmuA_v = qmu(muTest, asimov_data, pdf, init_pars, par_bounds)
 
     try:
```

Comment:
@lukasheinrich If I remember right, you had suggested that for the blocks below we always `clip` and not test first. Is this correct, or did I misunderstand you?

Reply:
Hi, yes the try except only works for the eager backends anyways (e.g. not TF), so I think we can just make this mandatory, and separately thinkg about how to deal with problematic situations